Polar errcaps

deep-jkl opened the pull request:
## PR Summary
Trying to reopen #21006, including fix for theta error bars.

I had to squash original contribution from @dstansby to pass through `PR cleanliness`, if there is any other way to reopen the original PR I would gladly do it (however, I couldn't figure it out).
The changes are minor, just changing the order of resolving "polar" criterion.
The fix is based on fact that to achieve markers perpendicular to the error bar, we need to find the error bar "midpoint theta", this is done by averaging lower and upper theta limit.

I have checked that it works for uneven error bars, see images comments.

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->
**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [x] New features are documented, with examples if plot related.
- [x] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [x] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of main, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

deep-jkl commented:
```python
import matplotlib.pyplot as plt

import numpy as np
fig = plt.figure(figsize=(10,10))
ax = plt.subplot(111, projection='polar')
theta = np.arange(0, 2*np.pi, np.pi / 4)
r = theta / np.pi / 2 + 0.5
ax.errorbar(theta, r, xerr=[0.15], xuplims=True, capsize=7)
```
![theta_errorbars_uplims](https://user-images.githubusercontent.com/48711526/184013966-8034bd77-af6f-4da4-bc6b-8c5c04e2dba9.png)

deep-jkl commented:
```python
import matplotlib.pyplot as plt

import numpy as np
fig = plt.figure(figsize=(10,10))
ax = plt.subplot(111, projection='polar')
theta = np.arange(0, 2*np.pi, np.pi / 4)
r = theta / np.pi / 2 + 0.5
xerr= np.vstack([[0.05]*8, [0.15]*8])
ax.errorbar(theta, r, xerr=xerr, capsize=7)
```
![theta_errorbars_uneven](https://user-images.githubusercontent.com/48711526/184014361-fab7b962-4acc-4a0c-9460-ed0c1f80a6be.png)

timhoffm commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -3600,6 +3600,16 @@ def test_errorbar():
+    ax = plt.subplot(111, projection='polar')
+    theta = np.arange(0, 2*np.pi, np.pi / 4)
+    r = theta / np.pi / 2 + 0.5
+    ax.errorbar(theta, r, xerr=0.15, yerr=0.1)
```
```suggestion
    ax.errorbar(theta, r, xerr=0.15, yerr=0.1, fmt='o')
```
Having a straight lines connecting the data points is rather misleading in polar plots. Even though this is just a test, let's use markers instead.

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -3585,15 +3586,36 @@ def apply_mask(arrays, mask): return [array[mask] for array in arrays]
+                            lo.get_xdata(), lo.get_ydata(),
+                            hi.get_xdata(), hi.get_ydata()):
+                    # Rotate caps to be perpendicular to the error bars
+                    rotation = (lo_theta + hi_theta)/2
```
I suppose the rotation angle should not be centered between lo and hi, but rather be defined by the theta data value. - Otherwise, e.g. extending the low value would increasingly tilt the errorbar. 

t.b.c. This may already be wrong in the calculation of the bar direction itself, not only in the caps here.

jklymak commented:
I guess I'm confused.  Is this a theta error?  If so, it should be drawn as a curve and kept at the same radius as the data point?

oscargus commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -3585,15 +3586,36 @@ def apply_mask(arrays, mask): return [array[mask] for array in arrays]
+                    self.add_line(mlines.Line2D([hi_theta], [hi_r],
+                                  marker=ms, **eb_cap_style))
+        else:
+            for axis in caplines:
+                for l in caplines[axis]:
+                    self.add_line(l)
```
```suggestion
            for axis_lines in caplines:
                for l in axis_lines:
                    self.add_line(l)
```

oscargus commented:
> If so, it should be drawn as a curve and kept at the same radius as the data point?

That is probably the ideal solution, yes. Similar to the lines not directly connecting the points, but uses some sort of curve. However, considering the current output this PR is clearly a step in the right direction, actually making it possible to use error caps.

On the other hand, it shouldn't be impossible to use an Arc instead and orient the caps pointing to the center. But if noone is implementing that, I'd clearly go for this solution in the meantime.

(One can also think about if the caps for errorbars in the other direction should be rounded or not?)

timhoffm commented:
Yes, if we want to be exact, theta errors should be curved. However, personally, I would not be that pedantic. Errors should be reasonably small (or the visualization does not tell much anyway). And fore small errors, a straight theta-line is a reasonable approximation of the curve. So if somebody wants to implement curved theta bars, great. But I would live with straight that-bars, if that's all we can do with reasonable effort now.

That said, I can also see an argument for not supporting theta bars at all if they are not curved.

jklymak commented:
If you can see the tilt for uncentred bars, I assume you can also see that they are not curved?

deep-jkl commented:
The original (rejected) solution in PR #21006 looked like this:
```
fig = plt.figure()
ax = plt.subplot(111, projection='polar')
theta = np.arange(0, 2*np.pi, np.pi / 4)
r = theta / np.pi / 2 + 0.5
ax.errorbar(theta, r, xerr=0.15, yerr=0.1)
ax.set_rlim(0, 2)
```
![image](https://user-images.githubusercontent.com/48711526/184247876-6af66918-17be-4383-8b96-cbc80befa33e.png)

On the other hand, under different settings I believe that curved theta error is must have, consider following plot:
```
import matplotlib.pyplot as plt

import numpy as np
fig = plt.figure(figsize=(10,10))
ax = plt.subplot(111, projection='polar')
theta = np.arange(0, 2*np.pi, np.pi / 4)
r = theta / np.pi / 2 + 0.5
ax.errorbar(theta, r, xerr=0.25, capsize=7, fmt="o")
```
![theta_big_bars](https://user-images.githubusercontent.com/48711526/184248551-07c30a1a-809c-4dd0-b9d7-fed9868d8a24.png)

timhoffm commented:
![grafik](https://user-images.githubusercontent.com/2836374/184251341-d40a2d35-3bec-4dbc-982e-44ef5968ae03.png)
Um, this is indeed awkward. Whether curved or not, the theta-error line should hit the data point.

Probably not the final call on polar error bars. Nevertheless, this PR fixes the caps for yerr and makes it a little less bad for xerr. So I approve as an improvement.



> (One can also think about if the caps for errorbars in the other direction should be rounded or not?)
IMHO this is not needed. Caps are essentially markers of the end point. They don't make a statement on the orthogonal axis. I.e. a y-cap does not say anything about theta. - Even if you disagress, from a practical point of view, caps are small enough so that the difference between rounded arc and straight line is small.

timhoffm commented:
Closes #441.

QuLogic commented:
> Closes #441.

This needs to go in the original description, a commit message, or you can link with the Description section in the right sidebar. Putting it here though, doesn't do anything. Unfortunately, #441 doesn't show up here to be linked, but if you go there, _this_ PR shows up to be linked, which I've now done.

deep-jkl commented:
I have fixed the testing figure. Additionally I have fixed the issue with radial offset for large error bars, see below:
```python
import matplotlib.pyplot as plt

import numpy as np
fig = plt.figure(figsize=(10,10))
ax = plt.subplot(111, projection='polar')
theta = np.arange(0, 2*np.pi, np.pi / 4)
r = theta / np.pi / 2 + 0.5
ax.errorbar(theta, r, xerr=0.25, capsize=7, fmt="o")
```
![nice_errorbars](https://user-images.githubusercontent.com/48711526/184534952-a0e86a51-efb2-4953-940b-01e68b737c2f.png)

Additionally I have done some experiments with curved error bars:
![curved_errorbars](https://user-images.githubusercontent.com/48711526/184534959-d9253229-619a-4bfb-8e40-16eebdc5164d.png)


You can see proof of principle with code changes [here](https://github.com/deep-jkl/matplotlib/tree/curved-polar-errcaps)

deep-jkl commented:
And just to make decision even harder, here I show some bar graphs with error bars:
```python
import matplotlib.pyplot as plt

import numpy as np
fig = plt.figure(figsize=(10,10))
ax = plt.subplot(111, projection='polar')
theta = np.arange(0, 2*np.pi, np.pi / 4)
r = theta / np.pi / 2 + 0.5
ax.errorbar(theta, r, xerr=0.25, yerr=0.1, capsize=7, fmt="o")
ax.bar(theta, r, width=0.4, bottom=0.0,  alpha=0.5)
```
Straight:
![polar_bars_straight](https://user-images.githubusercontent.com/48711526/184535269-ebfafb60-69f8-49e6-8014-cf2be7efb67c.png)

And curved:
![polar_bars_curved](https://user-images.githubusercontent.com/48711526/184535275-a720a906-48ed-4080-8eac-2ef42bbb232b.png)

oscargus commented:
Nice! I guess the best thing now is to discuss it at the dev-call on Thursday? Feel free to join if you want!

(Personally, I think that the curved error bars is the way to go.)

timhoffm commented:
I agree: The curved error bars is what we want.

tacaswell commented:
This was discussed on the weekly call.  The consensus was curved error bars and wait for it all to be ready for 3.7 rather than squeezing this in for 3.6 with straight lines.

We also discussed the interpolate step (private) feature of `Path` to simplify making the curved lines.

timhoffm commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -3600,6 +3600,16 @@ def test_errorbar():
     ax.set_title("Simplest errorbars, 0.2 in x, 0.4 in y")
 
 
+@image_comparison(['errorbar_polar_caps'], extensions=['png'])
```
```suggestion
@image_comparison(['errorbar_polar_caps'], extensions=['png'],
                  remove_text=True)
```

timhoffm commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -3600,6 +3600,16 @@ def test_errorbar():
+    ax = plt.subplot(111, projection='polar')
+    theta = np.arange(0, 2*np.pi, np.pi / 4)
+    r = theta / np.pi / 2 + 0.5
+    ax.errorbar(theta, r, xerr=0.15, yerr=0.1, fmt="o")
```
Maybe add additional datasets with:
- individual error bar sizes per data point
- asymmetric +/- errors
- a quite large theta-error (~pi/2 - this should still be curvy enough with 12 interpolation steps. I think it does not make sense to add a test for the case that the interpolation is not good anymore, because we may change this in the future - e.g. adapt the interpolation steps so that a step is never larger than a certain angle).

Can be 1 to few point per variant.

deep-jkl replied:
I have added more test cases, apart from the even errorbars defined by constant x/y err I mixed asymmetric and individual error bar sizes per datapoint. Additionally I have added long errorbars, and overlapping errorbars - these are not nice or useful, but on other hand it demonstrates that nothing goes particularly wrong.

With regards to _interpolation_steps, I changed the number to 2. It seems that it does work the same for all positive nubers greater than 1. I am not sure if this is desired behavior, but I didn't any deep investigation.

timhoffm replied:
Thanks! Could you join them into only one? We try quite hard to limit the amount of test images, because they blow up the size of the repo. The four tests would be an additional 200kb, which sounds not much, but we have thousands of tests, so it adds up.

Suggestion:
- outer most radius: one overlapping bar
- next inner radius: one large bar
- further in: one half for even bars, the other half for asymmetric bars.

If you plot them all into one Axes, automatic color cycling should give them different colors so that the different parts of the test are still easy to distinguish.

*Note*: I cannot comment on `_interpolation_steps` as that's not my topic, but emirically, it seems to work.

anntzer replied:
The fact that any _interpolation_steps > 1 works the same for fixed-radius arcs is intended, see https://github.com/matplotlib/matplotlib/pull/14852 (first paragraph of the first message).  Probably it could be better documented, though.

deep-jkl replied:
I have added the mix of errorbar options in single figure. But test failed for overlapping errorbars (see [this report](https://github.com/matplotlib/matplotlib/runs/7938210126?check_suite_focus=true) ). Therefore, for now, I have removed the overlapping case from the figure since it is not a real use case and its testing seems unstable.

timhoffm commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -3600,6 +3600,32 @@ def test_errorbar():
+def test_mixed_errorbar_polar_caps():
+    fig = plt.figure()
+    ax = plt.subplot(111, projection='polar')
+    theta = np.linspace(0, np.pi, 3, False)
+    r = [1.0]*3
+    ax.errorbar(theta, r, xerr=0.35, yerr=0.2, fmt="o")
+    _long_errorbar_polar_caps(ax)
+    _asymmetric_polar_caps(ax)
+
+
+def _long_errorbar_polar_caps(ax):
+    theta = np.linspace(0, 2 * np.pi, 2, False) + np.pi/2
+    r = [1.8, 2.2]
+    ax.errorbar(theta, r, xerr=0.8 * np.pi, yerr=0.2, fmt="o")
+
+
+def _asymmetric_polar_caps(ax):
+    theta = np.linspace(0, np.pi, 3, False) + np.pi
+    r = [1.0]*3
+    xerr = [[.3, .3, .2], [.2, .3, .3]]
+    yerr = [[.35, .5, .5], [.5, .35, .5]]
+    ax.errorbar(theta, r, xerr=xerr, yerr=yerr, fmt="o")
```
For symmetry and simplicity (less indirection) I'd put everything in one function:

```suggestion
    # symmetric errorbars
    theta = np.linspace(0, np.pi, 3, False)
    r = [1.0]*3
    ax.errorbar(theta, r, xerr=0.35, yerr=0.2, fmt="o")

    # long errorbars
    theta = np.linspace(0, 2 * np.pi, 2, False) + np.pi/2
    r = [1.8, 2.2]
    ax.errorbar(theta, r, xerr=0.8 * np.pi, yerr=0.2, fmt="o")

    # asymmetric errorbars
    theta = np.linspace(0, np.pi, 3, False) + np.pi
    r = [1.0]*3
    xerr = [[.3, .3, .2], [.2, .3, .3]]
    yerr = [[.35, .5, .5], [.5, .35, .5]]
    ax.errorbar(theta, r, xerr=xerr, yerr=yerr, fmt="o")
```

Additionally, I suggest not to position the bars on the grid. Pixel-wise interaction (antialiasing) between grid and errorbars could make the test less reproducible in the long run, and this is not something we want to test.
We can either

1. deactivate the grid (if that does not look too odd - the grid could be a visual help to verify for humans that the angles are right)
2. Slightly move the points in *r* and *theta*. - Note also that using explicit numbers may be simpler than `np.linspace`; e.g. for the long errorbar you could use `theta = [1.5, 3]`.

deep-jkl replied:
Ok, I have manually created some positions, and I have put overlapping back into the game. I have followed your advice and I have tuned the exact positions to avoid grid.
I added a comment to the test briefly summarizing the test purpose and hinting for future reference about the grid avoiding and possibility to turn them off if we arrive to problems.

jklymak reviewed (commented):
This looks good to me.  However, can it get an example for the gallery, and the what's-new link the example?

deep-jkl commented:
> This looks good to me. However, can it get an example for the gallery, and the what's-new link the example?

I have added an gallery example (just a simple one), but I am not sure that the link in `what's new` is valid.

jklymak commented on examples/pie_and_polar_charts/polar_error_caps.py:
```diff
@@ -0,0 +1,26 @@
+import matplotlib.pyplot as plt
+
+fig = plt.figure(figsize=(10, 10))
+ax = plt.subplot(111, projection='polar')
```
This would be more idiomatic, but not a big deal:
```suggestion
ax = fig.add_subplot(projection='polar')
```

deep-jkl replied:
I will fix it, I am taking some time to think of a more elaborate gallery example :-)

deep-jkl replied:
Ok, I could not think of something better. I have just added overlapping examples, including your suggestions.

jklymak commented on doc/users/next_whats_new/polar_errorbar_caps.rst:
```diff
@@ -0,0 +1,7 @@
+Caps and error lines are now drawn with respect to polar coordinates,
+when plotting errorbars on polar plots.
+
+.. figure:: /gallery/pie_and_polar_charts/images/sphx_glr_polar_error_caps_001.png
```
If this works, great!  (not sure it will, but my ignorance of sphinx is relatively vast)

deep-jkl replied:
I just mimicked it from [this](https://github.com/matplotlib/matplotlib/blob/main/doc/users/prev_whats_new/whats_new_3.5.0.rst#image-interpolation-now-possible-at-rgba-stage) what's new message, but it does not seem to work properly in the built documentation, so I am not sure it is worth the try.

jklymak reviewed (approved):
Small nit that you take or leave...
